source/use-cases/local-master-key.txt

```diff
@@ -0,0 +1,89 @@
+
+The **master key** is stored separately in a Key Management System (KMS) in order to prevent the **data encryption keys** from being used directly to access the sensitive data.
+
+The **data encryption keys** can safely be stored in a MongoDB collection alongside the encrypted sensitive data fields since they have been encrypted by the **master key**.
+
+In Client-Side Field Level Encryption, a client that has access to both **master key** and **data encryption keys** is necessary to perform encryption/decryption.
+
+In the code example below, we will set up the **master key** in a local KMS store as a file. This setup should only be used for local testing purposes and never for production. A secure service such as AWS KMS should always be used in production environments to ensure the **master key** is not compromised.
+
+
+Generate the local **master key** and save it to a file. You should only need to run this once:
+
+.. tabs::
+
+   tabs:
+
+     - id: java-master-key-generator
+       name: "Java"
+       content: |
+
+         .. code-block:: java
+
+           import java.io.FileOutputStream;
+           import java.io.IOException;
+           import java.security.SecureRandom;
+
+           public class CreateMasterKeyFile {
+             public static void main(final String[] args) {
+
+               final byte[] localMasterKey = new byte[96];
+               new SecureRandom().nextBytes(localMasterKey);
+
+               String path = "master-key.txt";
+               try (FileOutputStream stream = new FileOutputStream(path)) {
+                 stream.write(localMasterKey);
+               } catch (IOException e)  {
+                 e.printStackTrace();
+               }
+             }
+           }
+
+     - id: linux-macos-mater-key-generator
+       name: "Linux or MacOS"
+       content: |
+
+         .. code-block:: shell
+
+           echo $(head -c 64 /dev/urandom | base64 | tr -d '\n') >  'master-key.txt'
+
+     - id: windows-10plus-master-key-generator
+       name: "Windows 10+ Powershell"
+       content: |
+
+         .. code-block:: shell
+
+           powershell -command "[Convert]::ToBase64String((1..64|%{[byte](Get-Random -Max 256)}))" | Out-File -FilePath .\master-key.txt
+
+     - id: windows-pre10-master-key-generator
+       name: "Windows 8 and older Powershell"
+       content: |
+
+         .. code-block:: shell
+
+           powershell -command "$r=[byte[]]::new(64);$g=[System.Security.Cryptography.RandomNumberGenerator]::Create();$g.GetBytes($r);[Convert]::ToBase64String($r)" | Out-File -FilePath .\master-key.txt
+
+Next, in our MongoDB client application, add the following logic to read the **master key** from the file we created and construct the KMS settings:
+
+.. code-block:: java
```

Comment:
You need to add the tabs here, too, and have code snippets for each language, no?

Reply:
The current direction I think this page (open for debate) will take is that the entire guide (of which this is a single section) will be composed for a single language, and that a selector will allow the user to switch between the available languages. If the individual sections had language selectors, this could be confusing to the user when they get to the next code snippet and it is not in the same language that they had selected prior.

I think your comment surfaces a good point that providing multiple options could be confusing and instead we should steer the user towards doing everything one way. The prior tabbed block just shows users how to create a file which can be accomplished on command line or through the language's file management sdk.

Given the above explanation, what are your thoughts on whether to keep the prior code block? @MongoCaleb

Reply:
The tabs are designed so that, when you choose a language in the first example on the page, any other tabs on the page also switch to that language. That being said, it sounds like there may be a bigger picture to consider, and I'll defer to Ravind and/or Nick on that.

Reply:
Tabs are an area that we should do user research on, but for now let's think about these questions:
- Do users expect tabbed code blocks or do they think on a full-page level?
- Is the page a single piece of content (in this case probably yes) or will users follow links to specific sub-sections of this page (e.g. to view a specific example/step)

Given that this guide is more of a "top to bottom" flow, we could probably just use pills at the top and have non-tabbed code blocks. That said, I have an inkling that users will get confused so maybe we should also tab each code block individually to be safe.

For now, I don't think we need to have an answer to this - we can decide for the page holistically once the other pieces are compiled.

Reply:
Sounds like the addition of showing users how to create a local key file by just using command line tools would overcomplicate the UX. I'm satisfied with steering the user towards running a separate application to write the local key file, but would definitely want to discuss adding language-independent steps to language-specific guides at some point in the future.

Reply:
Removed the command line tool tabs